File: src/1-500/045/JumpGameII.py
from typing import List


class Solution:
    def jump(self, nums: List[int]) -> int:
        result = reach = maxReach = 0
        for index, num in enumerate(nums):
            if index > reach:
                reach = maxReach
                result += 1
            maxReach = max(maxReach, index + num)
        return result


solution = Solution()
print(solution.jump([0]))
print(solution.jump([2,3,1,1,4]))

File: src/1-500/045/JumpGameII.java
class Solution {
    public int jump(int[] nums) {
        if (nums == null || nums.length <= 1) {
            return 0;
        }
        int jump = 0;
        int curEnd = 0;
        int furthest = 0;
        for (int i = 0; i < nums.length; i++) {
            furthest = Math.max(furthest, i + nums[i]);
            if (curEnd == i) {
                jump++;
                curEnd = furthest;
            }
            if (curEnd >= nums.length - 1) {
                return jump;
            }
        }
        return jump;
    }
}